Add chained-comparison proposal (#10117)
dotnet/csharplang: proposals/chained-relational-comparison.md

## Motivation

Today, writing a range check in C# requires the programmer to mention the middle
operand twice:

```csharp
if (min <= x && x <= max) { … }
if (0 <= i && i < array.Length) { … }
```

This is awkward in three distinct ways:

**It does not match mathematical notation.** Students commonly learn to write
`min ≤ x ≤ max`.

**It is a double-evaluation correctness trap.** The natural rewrite of
`Min < ComputeVal() < Max` as `Min < ComputeVal() && ComputeVal() < Max`
evaluates `ComputeVal()` *twice*. That is wrong when the middle expression has
side-effects, is non-deterministic, or is merely expensive. The typical
workaround (binding the middle expression into a local), destroys the
expression-level naturalness of the condition:

```csharp
// What the user wants to write:
if (Min < ComputeVal() < Max) { … }

// What the user has to write today to avoid double-evaluation:
var tmp = ComputeVal();
if (Min < tmp && tmp < Max) { … }
```

**It is a copy-paste hazard.** The two-comparison form `a < b && c < d` reads
naturally as a chained comparison at a glance, but silently fails to test
a chain if `c` was meant to be `b`. There is no shorthand that makes the

> correct spelling the *only* spelling.

Reviewer: I get the point here, but I don't see what it has to do with copy-paste.

Also, in general, this type of substitution error is just one example of a far wider class of substitution errors which are not typically solved by introducing new syntax. Introducing new syntax for just this one case falls a bit short of being motivation for the feature personally. I see other benefits, but "moves the needle 1/10000th on substitution errors" isn't it (for me).

Author: Fair. This is just a restatement that a duplication can lead to bugs. I'll trim.

- **CoffeeScript**: [Chained comparisons](https://coffeescript.org/#comparisons).
- **C++26 (proposed)**: [P0893 — Chained comparisons](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0893r1.html).
- **Rust** explicitly reserved this syntactic space for a future chained-comparison feature: [RFC 558](https://rust-lang.github.io/rfcs/0558-require-parentheses-for-chained-comparisons.html) (accepted in 2015) banned the unparenthesised form `a < b < c` at compile time specifically to leave room for later adding Python-style chaining. [RFC issue #2083](https://github.com/rust-lang/rfcs/issues/2083) tracks the open "allow chaining of comparisons" proposal, and the third-party [`cmpchain`](https://docs.rs/cmpchain/latest/cmpchain/macro.chain.html) crate provides the feature via a macro today.

There is even a third-party NuGet workaround (referenced in the linked
discussion thread for this proposal) that uses operator-overload trickery to
approximate the feature, which is evidence of real demand in the ecosystem.

## Detailed design

The following updates are presented as a diff against the corresponding
sections of the C# 6 standard

> ([expressions.md](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md)).

Reviewer: Same comment as in other PR: would prefer draft-v8, or at least standard-v7

Throughout this section, ~~strikethrough~~ indicates text being removed from
the existing specification, and **bold** indicates text being added. Unchanged
prose is quoted verbatim for context.

### Grammar

No grammar change. The existing production for *relational_expression* in
[§11.11.1](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md#11111-general)
already parses `a < b < c < …` left-associatively as
`((a < b) < c) < …`. This proposal changes only how such a tree is bound and

> evaluated.

Reviewer: Can you clarify somewhere that `(a < b) < c` (with the explicit parenthesized_expression) is not proposed to behave as `a < b && b < c`?

Author: Absolutely.  It does fall out. But I had the same thought in the bus that we should be explicit here. As sometimes there use a view that parens don't change anything. And here it should.

is applied to select a specific operator implementation. ..."*:

**A *relational_expression* whose top-level operator is one of `<`, `<=`, `>`,
or `>=`, and whose left operand is itself a *relational_expression* whose
top-level operator is one of `<`, `<=`, `>`, or `>=`, may be a *chained
relational comparison*, bound per
[§11.11.13](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md#111113-chained-relational-comparison).
Because *relational_expression* is left-associative, the chain structure is
implicit in the parse tree; the rule in §11.11.13 applies recursively at each
such node.**

### §11.11.13 Chained relational comparison (new subsection)

Add the following new subsection at the end of
[§11.11](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md#1111-relational-and-type-testing-operators),
after [§11.11.12 The as operator](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md#111112-the-as-operator):

**This subclause defines the binding and run-time evaluation of
*relational_expression*s whose top-level operator is one of `<`, `<=`, `>`, or

> `>=` and whose left operand is itself such a *relational_expression*. Because

Reviewer: Is this worded strictly enough to not be applying to `((a | b) < c) < d`?
- The top level operator is `<` (from `< d`)
- the left operand is `(a | b) < c`, which is itself a relational expression due to the `<` in `< c`

Author: My upcoming edit will make this clear. This is not ready for review :-)

# Chained relational comparison

* Championed issue: <https://github.com/dotnet/csharplang/issues/8861>
* Discussion: <https://github.com/dotnet/csharplang/discussions/8643>

## Summary

Allow *relational_expression*s such as `a < b < c`, `min <= x <= max`, and
`0 <= i < array.Length` to have the intuitive chained-comparison meaning — roughly
`a < b && b < c`, `min <= x && x <= max`, `0 <= i && i < array.Length` —
without requiring the middle operand to be spelled twice. This extends naturally
to `<`, `<=`, `>`, `>=` combined in any order and at any length.

## Motivation

Today, writing a range check in C# requires the programmer to mention the middle
operand twice:

```csharp
if (min <= x && x <= max) { … }
if (0 <= i && i < array.Length) { … }
```

This is awkward in three distinct ways:

> **It does not match mathematical notation.** Every student learns to write

Author: ```suggestion
**It does not match mathematical notation.** Students commonly learn to write
```

`min ≤ x ≤ max`. C# is the only place the same inequality needs to be written

> as two separate comparisons joined by `&&`, with the pivot repeated.

Author: ```suggestion
`min ≤ x ≤ max`.
```

**It is a double-evaluation correctness trap.** The natural rewrite of
`Min < ComputeVal() < Max` as `Min < ComputeVal() && ComputeVal() < Max`
evaluates `ComputeVal()` *twice*. That is wrong when the middle expression has
side-effects, is non-deterministic, or is merely expensive. The typical

> workaround — binding the middle expression into a local — destroys the

Author: ```suggestion
workaround (binding the middle expression into a local), destroys the
```

## Summary

Allow *relational_expression*s such as `a < b < c`, `min <= x <= max`, and
`0 <= i < array.Length` to have the intuitive chained-comparison meaning,
without requiring the middle operand to be spelled twice. The semantics are

> roughly those of `a < b && b < c`, `min <= x && x <= max`, and

Reviewer: Instead of "roughly" which leaves things open and vague, maybe it could be "the semantics are those of" with the explicit exception that the middle operand is evaluated only once? (Or is there more difference than that?)
